File: 타겟넘버/java/20230120/src/Solution.java
class Solution {
    int answer = 0;

    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, 0);

        return answer;
    }
    public void dfs(int[] numbers, int target, int depth, int sum){
        if (depth == numbers.length){ //마지막 노드까지 진행했을 때
            if (target == sum){ //target 값과 합계가 같다면
                answer++;
            }

            return;
        }

        int plus = sum + numbers[depth]; //양수를 더한 값
        int minus = sum - numbers[depth]; //음수를 더한 값

        dfs(numbers, target, depth+1, plus);
        dfs(numbers, target, depth+1, minus);
    }
}

File: 타겟넘버/javascript/20230120/solution.js
function solution(numbers, target) {
    let answer = 0  //타겟 넘버를 만드는 방법의 수
    
    function dfs(idx, sum) {  //numbers의 인덱스와 현재까지의 합계
        if(idx==numbers.length) {  //numbers의 인덱스를 모두 탐색했다면
            if(sum==target) answer++;  //현재까지의 합계가 타겟 넘버이면 answer++
            
            return;
        }
        
        //아직 numbers를 끝까지 탐색하지 않았다면
        //해당 숫자를 더하거나 빼는 두 방법으로 나아간다
        dfs(idx+1, sum+numbers[idx]);  //해당 숫자를 더한다
        dfs(idx+1, sum-numbers[idx]);  //해당 숫자를 뺀다
    }
    
    dfs(0,0)  //함수 호출 (0번째 숫자, 현재까지 합계 0)
    
    return answer;
}
